Check for selected target in Alert wizard

dtedder opened the pull request:
added a property for tracking the selected target in the alert create wizard for spatial
created a method to get the single element returned in an identify results geoElements list added logic to check for the selected target in the click event of AlertConditionController added reset to -1 for the selected target index to the qml

related to #391

dtedder committed:
added a property for tracking the selected target in the alert create wizard for spatial
created a method to get the single element returned in an identify results geoElements list
added logic to check for the selected target in the click event of AlertConditionController
added reset to -1 for the selected target index to the qml

JamesMBallard commented on Shared/alerts/AlertConditionsController.h:
```diff
@@ -109,6 +110,7 @@ private slots:
   QJsonObject conditionToJson(AlertCondition* condition) const;
   QFuture<bool> addConditionFromJson(const QJsonObject& json);
   void addStoredConditions();
+  const std::optional<qint64> getPickedElementId(const QList<Esri::ArcGISRuntime::GeoElement*>& geoElements) const;
```
Nice, clean API 👍

JamesMBallard commented on Shared/alerts/AlertConditionsController.cpp:
```diff
@@ -710,6 +710,58 @@ void AlertConditionsController::onLayersChanged()
+  // is no longer the active tool or there were no geoelements returned
+  if (!isActive())
+    return {};
+  if (geoElements.count() < 1)
```
isEmpty is typical idiomatic Qt for checking containers for emptiness.

dtedder replied:
Yep, will swap this out.

JamesMBallard commented on Shared/alerts/AlertConditionsController.cpp:
```diff
@@ -710,6 +710,58 @@ void AlertConditionsController::onLayersChanged()
+    if (!attributes->containsAttribute(primaryKeyField))
+      return {};
+
+    return attributes->attributeValue(primaryKeyField).toInt();
```
On all code paths where you return an actual value, I think `std::make_optional` or just `std::optional{ ... }` is more explicit. My two cents would be to add something like that to make it clearer to the reader that's what is really happening. Without it, it is an implicit conversion.

dtedder replied:
Agreed, I'll make the update for the returns. Maybe also `std::nullopt` instead of the `{}` as well when there's nothing to return?

JamesMBallard replied:
>Maybe also std::nullopt

I will leave that up to you. IMO that's more subjective. Whenever you return `{}`, the reader can't jump to conclusions about what the type is, so it's less confusing. I do think returning `std::nullopt` is better overall.

JamesMBallard commented on Shared/alerts/AlertConditionsController.cpp:
```diff
@@ -710,6 +710,58 @@ void AlertConditionsController::onLayersChanged()
+    if (!attributes->containsAttribute(primaryKeyField))
+      return {};
+
+    return attributes->attributeValue(primaryKeyField).toInt();
```
Are there any warnings related to converting `toInt` to `qint64`?

dtedder replied:
Was not getting a warning here for any type conversion. However I am pretty sure this should be protected with `canConvert` to make sure it's not null from the attribute table.

JamesMBallard commented on Shared/alerts/AlertConditionsController.h:
```diff
@@ -144,6 +146,7 @@ private slots:
   QHash<QString,QString> m_messageFeedTypesToNames;
 
   QMetaObject::Connection m_mouseClickConnection;
+  qint16 m_selectedTargetNameIndex = -1;
```
If this is being passed in from QML land (I see the method is `Q_INVOKABLE`), then I would consider another data type. 

These are your choices, so I'd probably go with int.
https://doc.qt.io/archives/qt-5.15/qtqml-typesystem-basictypes.html#basic-types-provided-by-the-qml-language

QML has no support for `qint16` so you're truncating info from the caller, which is probably fine but not worth dealing with any unintended consequences (which would probably not happen, but why risk it).

dtedder replied:
Understood. Will switch to just `int`.

dtedder committed:
updates from PR draft discussion
- switched to plain int for QML interaction with CPP
- use .isEmpty for clearer check in condition on list
- use std::optional explicitly for return values
- guard QVariant value conversion with `canConvert`
- added missing forward declare for GeoElement in .h

dtedder committed:
missed the re-type to int on the Q_INVOKABLE method for the qml

dtedder commented:
@JamesMBallard I pushed another commit just now. Nothing big, just forgot to also update the type to plain `int` on the invokable method decl/def.

1f2f485618099fb1a4eedaf18d8f65a03f759382

dtedder committed:
- prevent the currently selected target layer from being de-selected if no elements were found on the map at the user's click location; just empty the <id> box
- clear the <id> box if the user selects another target layer from the drop-down list; prevent mismatched ids from previous clicks on map

dtedder commented:
@JamesMBallard another small cleanup based on testing. Clicking on the map where nothing exists will de-select the layer the user picked in the drop-down. Also, switching target layers should clear any previous ID to invalidate the form so the user can't click next and mistakenly apply the ID to a different target layer.

JamesMBallard reviewed (approved):
Latest changes 👍
